packages/core/test/odata/v4/entity.spec.ts

```diff
@@ -0,0 +1,93 @@
+import {
+  TestEntity,
+  TestEntityCircularLinkChild,
+  TestEntityCircularLinkParent,
+  TestEntityLvl2SingleLink,
+  TestEntityMultiLink,
+  TestEntitySingleLink
+} from '../../test-util/test-services/v4/test-service';
+
+describe('entity v4', () => {
+  let entity;
+
+  const state = {
+    stringProperty: '1',
+    collectionProperty: ['a', 'b'],
+    toSingleLink: {
+      toSingleLink: {
+        stringProperty: 'test'
+      },
+      customField: 2
+    },
+    toMultiLink: [
+      {
+        booleanProperty: false,
+        customField: 3
+      }
+    ],
+    customField: 1
+  };
+
+  beforeEach(() => {
+    entity = TestEntity.builder()
+      .stringProperty('1')
+      .collectionProperty(['a', 'b'])
+      .toSingleLink(
+        TestEntitySingleLink.builder()
+          .toSingleLink(
+            TestEntityLvl2SingleLink.builder().stringProperty('test').build()
+          )
+          .withCustomFields({
+            customField: 2
+          })
+          .build()
+      )
+      .toMultiLink([
+        TestEntityMultiLink.builder()
+          .booleanProperty(false)
+          .withCustomFields({
+            customField: 3
+          })
+          .build()
+      ])
+      .withCustomFields({
+        customField: 1
+      })
+      .build();
+  });
+  it('getCurrentMapKeys', () => {
+    const currentState = entity.getCurrentMapKeys();
+
+    expect(currentState).toEqual(state);
+    expect(currentState.collectionProperty).not.toBe(entity.collectionProperty);
+  });
+
+  it('setOrInitializeRemoteState() sets specific state', () => {
+    expect(entity.getUpdatedProperties()).toEqual(state);
+
+    entity.setOrInitializeRemoteState(state);
+
+    expect(entity.getUpdatedProperties()).toEqual({});
+  });
+
+  it('getCurrentMapKeys does not run endlessly', () => {
```

Comment:
can you add two things to be even more sure that the visited work:
- call `child['getCurrentMapsKeys'] so that we know that it does not depend on the direction
- make an example with a 1:N relation and also call it there from parent and one of the children.